### ISSUE TITLE
Add option to ignore differences in line endings during string comparison, so tests are not OS dependent

### ISSUE THREAD

**Issue by aSemy** [kotest/kotest#3161]
When writing tests that verify multi-line strings, for example file contents, it's common to use Kotest string assertions. However these tests are highly sensitive to line endings.

This assertion will pass on Unix, and fail on Windows:

https://github.com/kotest/kotest/blob/a7195a4d1c4a0d9e20526bc336338d394e028003/kotest-framework/kotest-framework-multiplatform-plugin-gradle/src/test/kotlin/io/kotest/framework/multiplatform/gradle/KotestMultiplatformCompilerGradlePluginSpec.kt#L38-L43

This PR also fixed line-ending tests failing https://github.com/kotest/kotest/pull/2743

I have some thoughts on how to implement this.

### Additional arg on matchers

One option is to add an arg, like `ignoreLineEndings`, to *all* string comparison matchers which defaults to true, but this would mean they could no longer be `infix fun`s. Adding an overload with a `ignoreLineEndings` arg could work around that, but that's a lot of work, code, maintenance, and testing.

> Example: new overloads are verbose and repetitive
> 
> ```kotlin
> // existing function
> infix fun <A : CharSequence?> A.shouldStartWith(prefix: CharSequence): A {
>    this should startWith(prefix)
>    return this
> }
> 
> // new overload
> fun <A : CharSequence?> A.shouldStartWith(prefix: CharSequence, ignoreLineEndings: Boolean = false): A {
>    this shouldNot startWith(prefix, ignoreLineEndings)
>    return this
> }
> 
> // existing function
> infix fun <A : CharSequence?> A.shouldNotStartWith(prefix: CharSequence): A {
>    this shouldNot startWith(prefix)
>    return this
> }
> 
> // new overload
> fun <A : CharSequence?> A.shouldNotStartWith(prefix: CharSequence, ignoreLineEndings: Boolean = false): A {
>    this shouldNot startWith(prefix, ignoreLineEndings)
>    return this
> }
> ```

### Boolean vs Sealed class flag

Even if there is an additional arg, there are drawbacks to using a boolean flag, as it is restricted in allowing for different options. Instead a sealed class/enum parameter would allow for more customisation, for example controlling exactly how the line endings are handled. Perhaps someone wants line endings in the 'expected' string to be converted from `/r/n` to `/n`, but on the 'actual' string left alone. Or vice versa, or both. 

### Project config option

A sealed class flag could be added to all string comparators, but it would be very verbose and repetitively. I think that all strings in a project would be compared the same way, there's no need for a per-assertion config, and a flag only needs to be set once.

One propose adding a [project configuration option](https://kotest.io/docs/framework/project-config.html) to control how line endings are dealt with.

```kotlin
object KotestProjectConfig : AbstractProjectConfig() {
    override val stringLineEndingsMode = StringLineEndings.MapToUnix
}
```

This flag can then be used in all string comparison matchers (but I'm not sure how this can be done... do the matchers have access to the project config?)

### Helper functions

Possibly the easiest option is just providing some helper functions, although their use in tests could be pretty repetitive and verbose

```kotlin
// note: this is just a quick example, the regex isn't correct
private val windows = Regex("\r\n")
private val unix = Regex("\n")

fun String.lineEndingsToUnix() = replace(windows, "\n")
fun String.lineEndingsToWindows() = replace(unix, "\r\n")
```

**Comment by Kantis** [comment]
I'm not sure if this is something Kotest should be concerned with?

In our case (with the gradle plugin spec), I think it's a [gitattribute](https://github.com/kotest/kotest/blob/master/.gitattributes#L1) which forces LF line-endings even on Windows. Just having auto would probably have avoided this?
https://github.com/kotest/kotest/blob/master/.gitattributes#L1

btw, how did you manage to inline that nice preview of the gradle spec? :)

**Comment by aSemy** [comment]
Fair enough - it's not required, and there are workarounds. It's just annoying when tests depend on the machine they're run on, and the original author might not realise until later.

Other libraries provide workarounds https://github.com/assertj/assertj/issues/818

In the example test I provided, the file `testReportFile` isn't checked into Git - it's a Gradle report xml under the build-directory - so Git can't fix it.


P.S. file preview instructions are here :) https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-a-permanent-link-to-a-code-snippet

**Comment by Kantis** [comment]
> Fair enough - it's not required, and there are workarounds. It's just annoying when tests depend on the machine they're run on, and the original author might not realise until later.

I agree, that should not be the case. I wonder why our Windows CI builds hasn't detected this failure. 🤔 

> In the example test I provided, the file `testReportFile` isn't checked into Git - it's a Gradle report xml under the build-directory - so Git can't fix it.

I got that. I think that Gradle realizes it's running on Windows and uses CRLF (`actual`) while Git has forced your checked out code to use LF (where the `expected` is defined). Using `auto` should check out the code to use CRLF on windows, and convert it to LF when checking in.

If we go back to discussing solutions, I would personally prefer to have the helper methods. But perhaps it would be good to leave the issue pinned for a while to gather more feedback?

> P.S. file preview instructions are here :) https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-a-permanent-link-to-a-code-snippet

Thanks! 😄

**Comment by aSemy** [comment]
> I got that. I think that Gradle realizes it's running on Windows and uses CRLF (`actual`) while Git has forced your checked out code to use LF (where the `expected` is defined). Using `auto` should check out the code to use CRLF on windows, and convert it to LF when checking in.

Ah yes, that explains it!

> If we go back to discussing solutions, I would personally prefer to have the helper methods. But perhaps it would be good to leave the issue pinned for a while to gather more feedback?

Sounds good!

Here's another idea: use a 'builder' style function. 

```kotlin
// receiver for the builder option, user can define options
interface StringMatchOptions {
   var ignoreLineEndings: Boolean
}

// create an overload that uses the options as a reciever
infix fun <A : CharSequence?> A.shouldStartWith(build: StringMatchOptions.() -> CharSequence): A {
   val options = object : StringMatchOptions {
      override var ignoreLineEndings: Boolean = false
   }
   val prefix = options.build()
   this should startWith(prefix, options)
   return this
}
```

The example usage is pretty nice:

```kotlin
fun test() {
   "aaa \n bbb \n ccc" shouldStartWith {
      ignoreLineEndings = true
      
      "aaa \r\n bbb \r\n ccc"
   }
}
```

**Comment by sksamuel** [comment]
I like the global setting option because you're likely to want to set it in one place.

**Comment by sksamuel** [comment]
I like the last suggestion too as we could have multiple options like trim and case insensitive

**Comment by Kantis** [comment]
> I like the last suggestion too as we could have multiple options like trim and case insensitive

I like it too. Perhaps the pattern could also be re-used for other things that also takes optional configuration, such as:

```kotlin
fun test() {
   """{"id": 1, "age": 20}""" shouldEqualJson {
      fieldComparison = FieldComparison.Lenient
      propertyOrder = PropertyOrder.Strict
      
      """{"age": 20}"""
   }

  timestamp shouldBeCloseTo {
     window = 2.seconds
     now()
  }
}
```

**Comment by sksamuel** [comment]
Yeah I think this is a great pattern

**Comment by Kantis** [comment]
Re-opening since there was another user requesting this on Slack. https://kotlinlang.slack.com/archives/CT0G9SD7Z/p1739802458634659
